jujugui/static/gui/src/app/app.js

```diff
@@ -551,79 +620,57 @@ Browser: ${navigator.userAgent}`
     ReactDOM.unmountComponentAtNode(
       document.getElementById('post-deployment')
     );
-    next();
   }
 
-  _renderHeaderLogo() {
-    const userName = this.user.displayName;
-    const gisf = this.applicationConfig.gisf;
+  _generateHeaderLogo() {
+    const userName = this.props.user.displayName;
+    const gisf = this.props.applicationConfig.gisf;
     const homePath = gisf ? '/' :
-      this.state.generatePath({profile: userName});
-    ReactDOM.render(
-      <HeaderLogo
-        gisf={gisf}
-        homePath={homePath}
-        showProfile={initUtils.showProfile.bind(this, this._bound.changeState, userName)} />,
-      document.getElementById('header-logo'));
+      this.props.appState.generatePath({profile: userName});
+    return (
+      <div className="header-banner__logo"
+        id="header-logo" >
+        <HeaderLogo
+          gisf={gisf}
+          homePath={homePath}
+          showProfile={initUtils.showProfile.bind(this, this._bound.changeState, userName)} />
+      </div>);
   }
 
   /**
     Renders the Charmbrowser component to the page in the designated element.
     @param {Object} state - The app state.
     @param {Function} next - Call to continue dispatching.
   */
-  _renderCharmbrowser(state, next) {
-    const charmstore = this.charmstore;
+  _generateCharmbrowser() {
+    const state = this.props.appState.current;
+    if (!this._isSet(state.store) && !this._isSet(state.search)) {
+      return null;
+    }
```

Comment:
Instead of adding/removing components from the DOM, they are shown/hidden via the URL state and rendered or not as we do any other React conditions.